core/math/projection.cpp

```diff
@@ -313,6 +314,48 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
 	*this = *this * cm;
 }
 
+void Projection::set_oblique(real_t p_fovy_degrees, real_t p_aspect, real_t p_z_near, real_t p_z_far, Quaternion p_oblique_plane, bool p_flip_fov) {
+	if (p_flip_fov) {
+		p_fovy_degrees = get_fovy(p_fovy_degrees, 1.0 / p_aspect);
+	}
+
+	real_t sine, cotangent, deltaZ;
+	real_t radians = Math::deg_to_rad(p_fovy_degrees / 2.0);
+
+	deltaZ = p_z_far - p_z_near;
+	sine = Math::sin(radians);
+
+	if ((deltaZ == 0) || (sine == 0) || (p_aspect == 0)) {
+		return;
+	}
+	cotangent = Math::cos(radians) / sine;
+
+	set_identity();
+
+	columns[0][0] = cotangent / p_aspect;
+	columns[1][1] = cotangent;
+	columns[2][2] = -(p_z_far + p_z_near) / deltaZ;
+	columns[2][3] = -1;
+	columns[3][2] = -2 * p_z_near * p_z_far / deltaZ;
+	columns[3][3] = 0;
+
+	// Here goes oblique magic!
+	// Eric Lengyel Solution: http://terathon.com/code/oblique.html
+	Quaternion q;
+
+	q.x = (SIGN(p_oblique_plane.x) + columns[2][0]) / columns[0][0];
+	q.y = (SIGN(p_oblique_plane.y) + columns[2][1]) / columns[1][1];
+
+	q.z = -1.0F;
+	q.w = (1.0F + columns[2][2]) / columns[3][2];
+
+	Quaternion c = p_oblique_plane * (2.0F / p_oblique_plane.dot(q));
+	columns[0][2] = c.x;
+	columns[1][2] = c.y;
+	columns[2][2] = c.z + 1.0F;
+	columns[3][2] = c.w;
```

Comment:
I would avoid taking a shortcut here, in case we do want to use this in XR. The math is the same (you'll be subtracting 0 for x, y and w and subtracting 1 for z, just like in your code) but this way it can be more easily refactored for other projection types in the future.
```suggestion
	columns[0][2] = c.x - columns[0][3];
	columns[1][2] = c.y - columns[1][3];
	columns[2][2] = c.z - columns[2][3];
	columns[3][2] = c.w - columns[3][3];
```

Reply:
I wasn't taking any shortcuts. These are the exact steps laid out in Eric Lengyel's algorithm (https://terathon.com/code/oblique.html).

I don't know enough linear algebra to determine if your suggestion will observe the same results, but as I quoted in the top PR comment, Eric's paper suggests that this algorithm is correct for any projection matrix, which I assume means that it can be used confidently unchanged.

Reply:
The generalization is explained tersely here: https://aras-p.info/texts/obliqueortho.html
from what it sounds like, this correction enables the math to work for any projection matrix, not just perspective.